### DISCUSSION TITLE
Modifying schemas with ZodType

### DISCUSSION THREAD

**Discussion by laneme** [colinhacks/zod#3202] [Q&A] [answered]
Omit and Pick methods are very useful to edit a schema and return new one. But this is limited to `ZodObject`

```ts
type Crawl = { url: string }
const schema: ZodType<Crawl> = z.object({ url: z.string() })

schema.omit // Property omit does not exist on type ...
```
ZodType loses omit/pick functions and makes it difficult to postprocess a schema that is assigned as `ZodType`.

Is there certain workarounds i can use for this purpose?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
type Crawl = { url: string }
const schema = z.object( { url: z.string() } ) satisfies z.ZodType<Crawl>
schema.omit // works as expected
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by laneme** [reply]
No no no, I was wondering if its possible to do it for schemas that are already assigned as
```ts
const schema: ZodType<Type> = ...
```
Schemas that comes out of third party library/generators is an example. Modifying them becomes an issue.

**Reply by JacobWeisenburger** [reply, marked as answer]
please provide an example of this

**Reply by CuriouslyCory** [reply]
Generated by prisma-zod-generator:
```ts
const Schema: z.ZodType<Prisma.ToppingCreateInput> = z
  .object({
    id: z.string().optional(),
    name: z.string(),
    category: z.lazy(() => CategorySchema),
    approved: z.boolean().optional(),
    createdAt: z.coerce.date().optional(),
    updatedAt: z.coerce.date().optional(),
    createdBy: z.lazy(() => UserCreateNestedOneWithoutToppingInputObjectSchema),
    Pizza: z
      .lazy(() => PizzaCreateNestedManyWithoutToppingsInputObjectSchema)
      .optional(),
  })
  .strict();
```

Not sure if same use case as OP, I want to omit the "createdBy" field so I can use the server session to populate it instead, but ZodType doesn't have pick/omit.

**Reply by JacobWeisenburger** [reply]
Is this what you are looking for?

```ts
const Schema = z.object( {
    id: z.string().optional(),
    name: z.string(),
    category: z.lazy( () => CategorySchema ),
    approved: z.boolean().optional(),
    createdAt: z.coerce.date().optional(),
    updatedAt: z.coerce.date().optional(),
    createdBy: z.lazy( () => UserCreateNestedOneWithoutToppingInputObjectSchema ),
    Pizza: z
        .lazy( () => PizzaCreateNestedManyWithoutToppingsInputObjectSchema )
        .optional(),
} ).strict() satisfies z.ZodType<Prisma.ToppingCreateInput>
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by CuriouslyCory** [reply]
That was my first thought so I did update the generator to provide that output but for some reason I was running into type errors anyway. I haven't had time to chase down the specifics yet.
